### PR TITLE
Modified External EPG module and new Selector module

### PR DESCRIPTION
**Pull request by shrsr**
Modified: mso_schema_template_externalepg
- Can associate ANP with the External EPG

New module: mso_schema_template_external_epg_selector
- Adds selectors to External EPGs

### CODE REVIEW

File: plugins/module_utils/mso.py [hunk 1]
```diff
@@ -145,6 +145,14 @@ def mso_expression_spec():
     )
 
 
+def mso_expression_spec_ext_epg():
+    return dict(
+        type=dict(type='str', required=True, aliases=['IP']),
```
Comment (lhercot) [inline plugins/module_utils/mso.py:150]: type should be a choice list with only IP as an option for now. I don't think Aliases is necessary here.

File: plugins/module_utils/mso.py [hunk 2]
```diff
@@ -145,6 +145,14 @@ def mso_expression_spec():
...
+        operator=dict(type='str', required=True),
```
Comment (lhercot) [inline plugins/module_utils/mso.py:151]: same here, operator is also a choice list and is only equal in that list today.

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2020, Cindy Zhao (@cizhao) <cizhao@cisco.com>
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:4]: You can leave Cindy's name but don't forget to add your own ;) It is your work too.

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: mso_schema_template_external_epg_selector
+short_description: Manage External EPG selector in schema templates
+description:
+- Manage External EPG selector in schema templates on Cisco ACI Multi-Site.
+author:
+- Cindy Zhao (@cizhao)
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:21]: Same here.

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    description:
+    - Expressions associated to this selector.
+    type: list
+    suboptions:
+      type:
+        description:
+        - The name of the expression which in this case is always IP address.
+        required: true
+        type: str
+        aliases: [ IP ]
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:52]: choices not aliases as mentioned above.

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+      operator:
+        description:
+        - The operator associated with the expression which in this case is always equals.
+        required: true
+        type: str
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:57]: Should also be a choice

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+      value:
+        description:
+        - The value of the IP Address associated with the expression.
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:60]: IP Address / Subnet

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+    choices: [ absent, present, query ]
+    default: present
+seealso:
+- module: mso_schema_template_externalepg
+extends_documentation_fragment: cisco.mso.modules
+'''
+
+EXAMPLES = r'''
+- name: Add a selector to an External EPG
+  cisco.mso.mso_schema_template_anp_epg_selector:
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:77]: This needs to be updated for external_epg_selector

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    schema: Schema 1
+    template: Template 1
+    external_epg: extEPG 1
+    selector: selector_1
+    expressions:
+      - type: ipAddress
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:86]: this should just be a choice and either choice is 'ip_address' or just 'ip' but we should comply with snake case and change it in the logic before putting it in payload.

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+        operator: equals
+        value: 10.0.0.0
+    state: present
+  delegate_to: localhost
+
+- name: Remove a Selector
+  cisco.mso.mso_schema_template_anp_epg_selector:
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:93]: module name also need to be updated here

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    password: SomeSecretPassword
+    schema: Schema 1
+    template: Template 1
+    external_epg: extEPG 1
+    selector: selector_1
+    state: absent
+  delegate_to: localhost
+
+- name: Query a specific Selector
+  cisco.mso.mso_schema_template_anp_epg_selector:
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:105]: module name also need to be updated here

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    schema: Schema 1
+    template: Template 1
+    external_epg: extEPG 1
+    selector: selector_1
+    state: query
+  delegate_to: localhost
+  register: query_result
+
+- name: Query all Selectors
+  cisco.mso.mso_schema_template_anp_epg_selector:
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:118]: module name also need to be updated here

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+    # Get template
+    templates = [t.get('name') for t in schema_obj.get('templates')]
+    if template not in templates:
+        mso.fail_json(msg="Provided template '{template}' does not exist. Existing templates: {templates}".format(template=template,
+                                                                                                                  templates=', '.join(templates)))
+    template_idx = templates.index(template)
+
+    # Get External EPG
+    externalEpgs = [e.get('name') for e in schema_obj.get('templates')[template_idx]['externalEpgs']]
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:181]: use snake case for the externalEpgs variable

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+    if external_epg not in externalEpgs:
+        mso.fail_json(msg="Provided external epg '{externalEpg}' does not exist. Existing epgs: {externalEpgs}"
+                      .format(externalEpg=external_epg, externalEpgs=', '.join(externalEpgs)))
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:184]: Use snake case for externalEpg and externalEpgs so external_epg and external_epgs

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+        elif not mso.existing:
+            mso.fail_json(msg="Selector '{selector}' not found".format(selector=selector))
+        mso.exit_json()
+
+    selectors_path = '/templates/{0}/externalEpgs/{1}/selectors/-'.format(template, external_epg)
+    ops = []
+
+    mso.previous = mso.existing
+    if state == 'absent':
+        mso.existing = {}
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:206]: mso.sent = mso.existing = {}

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+        ops.append(dict(op='remove', path=selector_path))
+
+    elif state == 'present':
+        # Get expressions
+        all_expressions = []
+        if expressions:
+            for expression in expressions:
+                IP = expression.get('type')
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:214]: don't use IP but prefer ip or ip_addr or ip_address as ALL_CAPS variable name are reserved for ENUMs or non mutable variables.

File: plugins/modules/mso_schema_template_externalepg.py [hunk 17]
```diff
@@ -80,6 +80,26 @@
...
+         required: true
+         type: str
+       schema:
+         description:
+         - The schema that defines the referenced anp.
+         - If this parameter is unspecified, it defaults to the current schema.
+         type: str
+       template:
+         description:
+         - The template that defines the referenced L3Out.
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_externalepg.py:100]: referenced anp

File: plugins/module_utils/mso.py [hunk 18]
```diff
@@ -490,25 +498,30 @@ def vrf_dict_from_ref(self, data):
...
-            'contracts': ['contractName', 'schemaId', 'templateName'],
-            'l3outs': ['l3outName', 'schemaId', 'templateName']
-        }
-        result = {
-            uri_map[category][0]: name,
-            uri_map[category][1]: schema_id,
-            uri_map[category][2]: template_name,
-        }
-        return result
+        if data:
```
Comment (lhercot) [inline plugins/module_utils/mso.py:501]: if data and data != '':

File: plugins/module_utils/mso.py [hunk 19]
```diff
@@ -490,25 +498,30 @@ def vrf_dict_from_ref(self, data):
...
+            ref_regex = re.compile(r'\/schemas\/(.*)\/templates\/(.*)\/(.*)\/(.*)')
+            dic = ref_regex.search(data)
+            if dic:
```
Comment (lhercot) [inline plugins/module_utils/mso.py:504]: if dic and 'group' in dic:

File: plugins/modules/mso_schema_template_externalepg.py [hunk 20]
```diff
@@ -228,13 +253,15 @@ def main():
...
         l3out_ref = mso.make_reference(l3out, 'l3out', schema_id, template)
+        anp_ref = mso.make_reference(anp, 'anp', schema_id, template)
         if display_name is None and not mso.existing:
             display_name = externalepg
 
         payload = dict(
             name=externalepg,
             displayName=display_name,
             vrfRef=vrf_ref,
+            anpRef=anp_ref,
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_externalepg.py:264]: don't always add it to payload.
Remove it from here and add a line after payload like:
if anp is not None:
    payload['anpRef'] = anp_ref

File: plugins/module_utils/mso.py (hunk 19 above)
Comment (shrsr) [reply plugins/module_utils/mso.py:504]: Left it as-
If dic is not None:

We can make a change if we hit something down the road.

File: plugins/modules/mso_schema_template_externalepg.py (hunk 20 above)
Comment (shrsr) [reply plugins/modules/mso_schema_template_externalepg.py:264]: Done!

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+    ops = []
+
+    mso.previous = mso.existing
+    if state == 'absent':
+        mso.sent = mso.existing = {}
+        ops.append(dict(op='remove', path=selector_path))
+
+    elif state == 'present':
+        # Get expressions
+        all_expressions = []
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:214]: types = dict(ip_address='ipAddress')

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+        if expressions:
+            for expression in expressions:
+                ip_addr = expression.get('type')
+                if ip_addr is not None:
+                    ip_addr = 'ipAddress'
+                operator = expression.get('operator')
+                value = expression.get('value')
+                all_expressions.append(dict(
+                    key=ip_addr,
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:223]: key = types[type]

File: plugins/modules/mso_schema_template_external_epg_selector.py [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+    if state == 'absent':
+        mso.sent = mso.existing = {}
+        ops.append(dict(op='remove', path=selector_path))
+
+    elif state == 'present':
+        # Get expressions
+        all_expressions = []
+        if expressions:
+            for expression in expressions:
+                ip_addr = expression.get('type')
```
Comment (lhercot) [inline plugins/modules/mso_schema_template_external_epg_selector.py:217]: type = expression.get('type')

Comment (lhercot) [reply plugins/modules/mso_schema_template_external_epg_selector.py:217]: delete next two lines

**Pull request merged**